Add fallback for on demand policy

ZhaoxiangJin opened the pull request:
In current on_demand policy, if it doesn't find a P-state with a threshold lower than the current load, it will fail to switch state. For example, if load went from 100% -> 0% suddenly as the chip would get stuck in a "turbo mode" meanwhile its load is low.

This commit add fallback mechanism for on-demand policy, that is if no P-state matches (i.e., the load is below all thresholds), the policy will fall back to the P-state with the lowest load_threshold (typically the lowest performance state).

See https://github.com/zephyrproject-rtos/zephyr/pull/98427  for some related discussions.

ZhaoxiangJin committed:
cpu_freq: Add fallback mechanism for on-demand policy

In current on_demand policy, if it doesn't find a P-state
with a threshold lower than the current load, it will fail
to switch state. For example, if load went from 100% -> 0%
suddenly as the chip would get stuck in a "turbo mode"
meanwhile its load is low.

This commit add fallback mechanism for on-demand policy,
that is if no P-state matches (i.e., the load is below all
thresholds), the policy will fall back to the P-state with
the lowest load_threshold (typically the lowest performance
state).

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
doc: cpu_freq: add details about on-demand policy

add details about on-demand policy

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

seankyer commented on subsys/cpu_freq/policies/on_demand/on_demand.c:
```diff
@@ -36,47 +37,60 @@ static unsigned int num_unprocessed_cpus;
 {
-	int cpu_load;
-	int cpu_id = 0;
+	if (ARRAY_SIZE(soc_pstates) == 0) {
```
Am I mistaken? Or is this already checked in the core `cpu_freq.c`? If not, let's move it there. A board cannot use cpu_freq subsystem at all if it has no defined soc_pstates.

seankyer commented on doc/services/cpu_freq/policies/on_demand.rst:
```diff
@@ -7,8 +7,18 @@ The On-Demand policy evaluates the current CPU load using the
+The On-Demand policy will iterate through the defined P-states and select the first P-state for which
+the CPU load is greater than or equal to the defined threshold.
+
+**Fallback behavior:**
```
IMO, we don't need to call this anything special like "fallback". Just state that it is an intrinsic behavior of this policy that if the user doesn't define a 0 threshold, the lowest soc pstate will be used. The suggestion to explicitly set this to 0 is more of an implementation note and I don't think is relevant for the user docs.

seankyer reviewed (changes requested):
Thank you for the PR!
@kartben since this is a bugfix, is there a label we must add? Could go into 4.3?

ZhaoxiangJin committed:
cpu_freq: Validate P-states in devicetree in build time

Validate P-states in devicetree in build time. If no
P-states are defined, a build error will be raised.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin commented on subsys/cpu_freq/policies/on_demand/on_demand.c:
```diff
@@ -36,47 +37,60 @@ static unsigned int num_unprocessed_cpus;
+				i, state->load_threshold);
 			return 0;
 		}
+		if (state->load_threshold < min_threshold) {
```
good ideal, updated.

seankyer commented on doc/services/cpu_freq/policies/on_demand.rst:
```diff
@@ -7,8 +7,15 @@ The On-Demand policy evaluates the current CPU load using the
+the CPU load is greater than or equal to the defined threshold.
+
+
+**Intrinsic behavior:**
```
Nit: we don't need this header

seankyer reviewed (changes requested):
Almost there. Thank you. You have some failing tests it seems.

ZhaoxiangJin committed:
tests: on_demand: Add test for P-states order

Ensure that the P-states defined in devicetree for the CPU frequency
scaling on-demand policy are in decreasing order of load thresholds.
This is important because the on-demand policy relies on this order
to select the appropriate P-state based on current CPU load.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

seankyer commented on tests/subsys/cpu_freq/policies/on_demand/src/main.c:
```diff
@@ -15,6 +15,23 @@ LOG_MODULE_REGISTER(cpu_freq_on_demand_test, LOG_LEVEL_INF);
+ */
+ZTEST(cpu_freq_on_demand, test_pstates_order)
+{
+	zassert_true(soc_pstates_count > 1, "There must be more than one P-state defined");
```
Actually it's not true there must be more than 1, there must be _at least_ 1 for the subsystem to function.

seankyer reviewed (commented):
Couple last notes. Thanks for your patience.

seankyer reviewed (approved):
LGTM thanks

ZhaoxiangJin commented:
> @ZhaoxiangJin - if this is a bugfix that is intended to go into 4.3.0, can you please link a bug report issue to this PR?

It is a improvement, let's target 4.4.0
